kernel/cmdline.cc

```diff
@@ -93,33 +120,135 @@ static u64 parse_uint_option(const char* name, long default_value) {
   return (u64)ret;
 }
 
-static void parse_string_option(const char* name, const char *default_value, char *output) {
+static void parse_string_param(const char* name, const char *default_value, char *output) {
   if(!getvalue(name, output) || output[0] == '\0') {
     strncpy(output, default_value, CMDLINE_VALUE);
     output[CMDLINE_VALUE] = '\0'; // just in case every byte was filled
   }
 }
 
+void
+view_binary_param(struct param_metadata<bool> *param)
+{
+  cprintf("%s: %s (bool)\n", param->name, *param->pval ? "yes" : "no");
+}
+
+void
+view_uint_param(struct param_metadata<u64> *param)
+{
+  cprintf("%s: %lu (uint)\n", param->name, *param->pval);
+}
+
+void
+view_string_param(struct param_metadata<char *> *param)
+{
+  cprintf("%s: %s (str)\n", param->name, (char*) param->pval);
+}
+
+// print the value of the specified param, or all params if name is null
+int
+cmdline_view_param(const char *name)
+{
+  for (auto &param : binary_params) {
+    if (name == NULL || strcmp(param.name, name) == 0)
+      view_binary_param(&param);
+  }
+  for (auto &param : uint_params) {
+    if (name == NULL || strcmp(param.name, name) == 0)
+      view_uint_param(&param);
+  }
+  for (auto &param : string_params) {
+    if (name == NULL || strcmp(param.name, name) == 0)
+      view_string_param(&param);
+  }
+  return 0;
+}
+
+void
+pause_and_call(void (*fn)(void))
```

Comment:
Could you move this function to ipi.cc and make it the only exported interface to the pause functionality? All this pausing logic is somewhat scary since it could deadlock the system if we aren't careful. I'd like to move it all to once place as much as possible.